### ISSUE TITLE
os.getAppFilename does not handle edge case on Linux

### ISSUE THREAD

**Issue by iacore** [nim-works/nimskull#726]
tl;dr: `/proc/self/exe` is not reliable. Use `argv[0]` or `fork(2)` instead.

> Under Linux 2.2 and later, ... If the pathname
              has  been  unlinked,  the  symbolic  link will contain the
              string '(deleted)' appended to the original pathname.   In
              a  multithreaded  process,  the  contents of this symbolic
              link are not available if the main thread has already ter‐
              minated (typically by calling pthread_exit(3)). -- [proc(5)](https://man7.org/linux/man-pages/man5/proc.5.html)

Here's an example
```nim
import std/os
while true:
  echo getAppFilename()
  sleep 1000
```

1. compile with `nim c xxx.nim`
2. run the executable
3. delete the executable while it's running

**Comment by alaviss** [comment]
`argv[0]` is even more unreliable, as any arbitrary value can be set by the caller.

I think we should just educate users about it instead, like how I did it in: https://github.com/nim-lang/Nim/pull/15093

**Comment by iacore** [comment]
> argv[0] is even more unreliable, as any arbitrary value can be set by the caller.

I think it is set by the kernel (`exec*` syscalls), and then processed by libc `_entry`.

> I think we should just educate users about it instead, like how I did it in: https://github.com/nim-lang/Nim/pull/15093

That's good!

**Comment by alaviss** [comment]
> I think it is set by the kernel (`exec*` syscalls), and then processed by libc `_entry`.

`execve(2)`'s `argv` parameter is the entire `argv` array that is passed to the callee. Naturally, this includes `argv[0]`.

If you use bash, you can set this to any arbitrary value using `exec -a <name> <command>`.

We should probably port that documentation PR over here, given how it was rejected in Nim.
